### PR TITLE
Add new redis database instance for BMP_STATE_DB.

### PR DESCRIPTION
**Pull request by FengPan-Frank**
#### Why I did it
Add new redis database instance for BMP_STATE_DB. since bmp data set will be a bit large which we don't want block existing state_db operation.

##### Work item tracking
- Microsoft ADO **(number only)**:27588893

#### How I did it
Changed database config and will use it into other bmp relevant PR.

#### How to verify it

image
image

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

File: dockers/docker-database/database_config.json.j2 [hunk 1]
```diff
@@ -130,6 +130,12 @@
             "instance" : {% if include_remote_db %} "remote_redis" {% else %} "redis" {% endif %}
         }
 {% endif %}
+        ,
+        "BMP_STATE_DB" : {
+            "id" : 19,
+            "separator": "|",
+            "instance" : "redis"
```
Comment (qiluo-msft) [inline dockers/docker-database/database_config.json.j2:137]: >redis [](http://example.com/codeflow?start=26&length=5)

This is still the single instance. Will you use another dedicated redis instance? It's okay to have multiple redis-server instances inside one database container.

Comment (FengPan-Frank) [reply dockers/docker-database/database_config.json.j2:143]: updated to use new redis instance as well.

Comment (qiluo-msft) [reply dockers/docker-database/database_config.json.j2:143]: Do you mean `redis_bmp` instance?

File: dockers/docker-database/database_config.json.j2 [hunk 2]
```diff
@@ -24,6 +24,12 @@
             "persistence_for_warm_boot" : "yes"
         }
 {% endif %}
+        ,"redis_bmp":{
+            "hostname" : "redis_bmp.server",
+            "port": 6399,
```
Comment (qiluo-msft) [inline dockers/docker-database/database_config.json.j2:29]: >6399 [](http://example.com/codeflow?start=20&length=4)

Magic number.
Actually I see the value of REMOTE_DB_PORT named constant. It help organized the port allocation in a central place, easy to detect port confliction. Could you follow?

File: dockers/docker-database/database_config.json.j2 (hunk 1 above)
Comment (FengPan-Frank) [reply dockers/docker-database/database_config.json.j2]: fixed.

File: dockers/docker-database/database_config.json.j2 (hunk 2 above)
Comment (FengPan-Frank) [reply dockers/docker-database/database_config.json.j2]: remove port since we only have local connection requirement, thus domain socket is fine.

**Review by qiluo-msft — DISMISSED** [review]

File: dockers/docker-database/database_config.json.j2 [hunk 3]
```diff
@@ -24,6 +24,12 @@
             "persistence_for_warm_boot" : "yes"
         }
 {% endif %}
+        ,"redis_bmp":{
+            "hostname" : "{{HOST_IP}}",
+            "port" : {{BMP_DB_PORT}},
+            "unix_socket_path": "/var/run/redis-bmp/redis_bmp.sock",
```
Comment (Pterosaur) [inline dockers/docker-database/database_config.json.j2:30]:
```suggestion
            "unix_socket_path": "/var/run/redis/redis_bmp.sock",
```

Comment (Pterosaur) [reply dockers/docker-database/database_config.json.j2:30]:
```
redis_bmp "# /bin/bash -c "{ [[ -s /var/lib/redis_bmp/dump.rdb ]] || rm -f /var/lib/redis_bmp/dump.rdb; } && mkdir -p /var/lib/redis_bmp && exec /usr/bin/redis-server /etc/redis/redis.conf --bind  127.0.0.1 --port 6400 --unixsocket /var/run/redis-bmp/redis_bmp.sock --pidfile /var/run/redis/redis_bmp.pid --dir /var/lib/redis_bmp "
542:C 31 May 2024 00:20:58.218 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
542:C 31 May 2024 00:20:58.218 # Redis version=7.0.15, bits=64, commit=00000000, modified=0, pid=542, just started
542:C 31 May 2024 00:20:58.218 # Configuration loaded
542:M 31 May 2024 00:20:58.218 * monotonic clock: POSIX clock_gettime
542:M 31 May 2024 00:20:58.219 # Failed opening Unix socket: bind: No such file or directory
```

Comment (FengPan-Frank) [reply dockers/docker-database/database_config.json.j2]: Thanks for the pointer, fixed.

File: dockers/docker-database/database_config.json.j2 [hunk 4]
```diff
@@ -130,6 +136,18 @@
             "instance" : {% if include_remote_db %} "remote_redis" {% else %} "redis" {% endif %}
         }
 {% endif %}
+        ,
+        "EVENT_DB" : {
```
Comment (qiluo-msft) [inline dockers/docker-database/database_config.json.j2:140]: >EVENT_DB [](http://example.com/codeflow?start=9&length=8)

Is it your code change?

Comment (qiluo-msft) [reply dockers/docker-database/database_config.json.j2:140]: @zbud-msft do you have any idea?

File: dockers/docker-database/docker-database-init.sh [hunk 5]
```diff
@@ -41,14 +41,16 @@ then
     fi
 fi
 
+export BMP_DB_PORT=6400
```
Comment (qiluo-msft) [inline dockers/docker-database/docker-database-init.sh:44]: >6400 [](http://example.com/codeflow?start=19&length=4)

@Pterosaur How to ensure this port is not conflicting with REMOTE_DB_PORT? or does it matter?

File: rules/docker-restapi.mk [hunk 6]
```diff
@@ -29,6 +29,7 @@ endif
 $(DOCKER_RESTAPI)_CONTAINER_NAME = restapi
 $(DOCKER_RESTAPI)_RUN_OPT += -t
 $(DOCKER_RESTAPI)_RUN_OPT += -v /var/run/redis/redis.sock:/var/run/redis/redis.sock
+$(DOCKER_RESTAPI)_RUN_OPT += -v /var/run/redis/redis_bmp.sock:/var/run/redis/redis_bmp.sock
```
Comment (qiluo-msft) [inline rules/docker-restapi.mk:32]: >DOCKER_RESTAPI [](http://example.com/codeflow?start=2&length=14)

why this feature is related to restapi?

File: platform/vs/docker-sonic-vs/database_config.json [hunk 7]
```diff
@@ -11,6 +11,11 @@
             "port": 6380,
             "unix_socket_path": "/var/run/redis/redis_chassis.sock"
         }
+        ,"redis_bmp":{
```
Comment (zbud-msft) [inline platform/vs/docker-sonic-vs/database_config.json:14]: nit:  Can we put comma at the end of the previous section?

File: platform/vs/docker-sonic-vs/database_config.json [hunk 8]
```diff
@@ -77,6 +82,11 @@
             "id" : 14,
             "separator": ":",
             "instance" : "redis"
+        },
+        "BMP_STATE_DB" : {
+            "id" : 20,
```
Comment (zbud-msft) [inline platform/vs/docker-sonic-vs/database_config.json:87]: Why is new database id 20 vs next in sequence 15?

File: platform/vs/docker-sonic-vs/database_config.json (hunk 7 above)
Comment (FengPan-Frank) [reply platform/vs/docker-sonic-vs/database_config.json:14]: done

File: dockers/docker-database/docker-database-init.sh (hunk 5 above)
Comment (Pterosaur) [reply dockers/docker-database/docker-database-init.sh:44]: Yes, it's possible. But according to the current smart-switch architecture, the 19 port quotas are sufficient.

File: rules/docker-restapi.mk (hunk 6 above)
Comment (FengPan-Frank) [reply rules/docker-restapi.mk]: removed unnecessary change.

File: platform/vs/docker-sonic-vs/database_config.json (hunk 8 above)
Comment (FengPan-Frank) [reply platform/vs/docker-sonic-vs/database_config.json]: removed unnecessary change

File: dockers/docker-database/supervisord.conf.j2 [hunk 9]
```diff
@@ -42,13 +42,13 @@ dependent_startup=true
 {%- set ADDITIONAL_OPTS = '' %}
 {%- endif -%}
 command=/bin/bash -c "{ [[ -s /var/lib/{{ redis_inst }}/dump.rdb ]] || rm -f /var/lib/{{ redis_inst }}/dump.rdb; } && mkdir -p /var/lib/{{ redis_inst }} && exec /usr/bin/redis-server /etc/redis/redis.conf --bind {{ LOOPBACK_IP }} {{ redis_items['hostname'] }} --port {{ redis_items['port'] }} --unixsocket {{ redis_items['unix_socket_path'] }} --pidfile /var/run/redis/{{ redis_inst }}.pid --dir /var/lib/{{ redis_inst }} {{ ADDITIONAL_OPTS }}"
-priority=2
```
Comment (qiluo-msft) [inline dockers/docker-database/supervisord.conf.j2:45]: Changes in this file seems not related to BMP feature. Why changing?

File: files/build_templates/docker_image_ctl.j2 [hunk 10]
```diff
@@ -90,10 +90,12 @@ function preStartAction()
         if [[ ("$BOOT_TYPE" == "warm" || "$BOOT_TYPE" == "fastfast" || "$BOOT_TYPE" == "fast")  && -f $WARM_DIR/dump.rdb ]]; then
             # Load redis content from /host/warmboot/dump.rdb
             docker cp $WARM_DIR/dump.rdb database$DEV:/var/lib/redis/dump.rdb
+            docker cp $WARM_DIR/dump.rdb database$DEV:/var/lib/redis_bmp/dump.rdb
```
Comment (qiluo-msft) [inline files/build_templates/docker_image_ctl.j2:93]: >docker cp $WARM_DIR/dump.rdb database$DEV:/var/lib/redis_bmp/dump.rdb [](http://example.com/codeflow?start=12&length=69)

BGP feature has no special handling for Warm restart support. The code change here will hurt warm bootup process. Why changing?

File: dockers/docker-database/supervisord.conf.j2 (hunk 9 above)
Comment (FengPan-Frank) [reply dockers/docker-database/supervisord.conf.j2]: Revert this back, we can keep redis and redis_bmp as the same priority.

File: files/build_templates/docker_image_ctl.j2 (hunk 10 above)
Comment (FengPan-Frank) [reply files/build_templates/docker_image_ctl.j2]: yes, we should not need this warm restart operation. removed.

File: dockers/docker-database/supervisord.conf.j2 (hunk 9 above)
Comment (qiluo-msft) [reply dockers/docker-database/supervisord.conf.j2]: Below change is also not related. could you just revert the whole file?

Comment (FengPan-Frank) [reply dockers/docker-database/supervisord.conf.j2]: This change is necessary, otherwise multiple db instance will generate invalid format of supervisord.conf without starting new line.

Comment (FengPan-Frank) [reply dockers/docker-database/supervisord.conf.j2]: revert this bug fix change, and use separate PR instead https://github.com/sonic-net/sonic-buildimage/pull/20545

**Review by xincunli-sonic — APPROVED** [review]
:shipit:

**Review by xincunli-sonic — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]
